File: 103. Binary Tree Zigzag Level Order Traversal/103. Binary Tree Zigzag Level Order Traversal.dart
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *   int val;
 *   TreeNode? left;
 *   TreeNode? right;
 *   TreeNode([this.val = 0, this.left, this.right]);
 * }
 */
class Solution {
    List<List<int>> zigzagLevelOrder(TreeNode? root) {
        if (root == null) return [];
        
        List<List<int>> result = [];
        List<TreeNode> queue = [root];
        bool leftToRight = true;
        
        while (queue.isNotEmpty) {
            int levelSize = queue.length;
            List<int> currentLevel = List.filled(levelSize, 0);
            
            for (int i = 0; i < levelSize; i++) {
                TreeNode node = queue.removeAt(0);
                int index = leftToRight ? i : levelSize - 1 - i;
                currentLevel[index] = node.val;
                
                if (node.left != null) queue.add(node.left!);
                if (node.right != null) queue.add(node.right!);
            }
            
            leftToRight = !leftToRight;
            result.add(currentLevel);
        }
        
        return result;
    }
}

File: 103. Binary Tree Zigzag Level Order Traversal/103. Binary Tree Zigzag Level Order Traversal.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public IList<IList<int>> ZigzagLevelOrder(TreeNode root) {
        IList<IList<int>> result = new List<IList<int>>();
        if (root == null) return result;
        
        Queue<TreeNode> queue = new Queue<TreeNode>();
        queue.Enqueue(root);
        bool leftToRight = true;
        
        while (queue.Count > 0) {
            int levelSize = queue.Count;
            int[] currentLevel = new int[levelSize];
            
            for (int i = 0; i < levelSize; i++) {
                TreeNode node = queue.Dequeue();
                int index = leftToRight ? i : levelSize - 1 - i;
                currentLevel[index] = node.val;
                
                if (node.left != null) queue.Enqueue(node.left);
                if (node.right != null) queue.Enqueue(node.right);
            }
            
            leftToRight = !leftToRight;
            result.Add(currentLevel.ToList());
        }
        
        return result;
    }
}